File: python/tests/test_ExtensionInjector.py
# MODCDP_TRANSLATE_TEST: KEEP THIS TEST FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
# All test cases, descriptions, covered edge cases, and setup should be kept perfectly 1:1 in sync between:
# - ./js/test/test.ExtensionInjector.ts
# - ./go/modcdp/injector/ExtensionInjector_test.go
# NO MOCKING, NO MONKEY PATCHING, NO SIMULATING, NO FAKING, NO SKIPPING ALLOWED.
# USE REAL USER-FACING CODE PATHS WITH REAL BROWSERS, REAL CLASSES, REAL URLS, etc. Hard fail if keys or other env requirements are missing.
from __future__ import annotations

import unittest

from modcdp.injector.ExtensionInjector import ExtensionInjector


class ExtensionInjectorTests(unittest.TestCase):
    def test_extensioninjector_owns_shared_injector_config(self) -> None:
        injector = ExtensionInjector(
            {
                "injector_service_worker_extension_id": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
                "injector_service_worker_url_suffixes": ["/modcdp/service_worker.js"],
            }
        )

        self.assertEqual(injector.config.injector_service_worker_extension_id, "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
        self.assertEqual(injector.configForUpstream(), {})
        self.assertEqual(injector.extra_args, [])
        self.assertTrue(
            injector._serviceWorkerTargetMatches(
                {
                    "targetId": "target-1",
                    "type": "service_worker",
                    "url": "chrome-extension://aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa/modcdp/service_worker.js",
                }
            )
        )
        self.assertFalse(
            injector._serviceWorkerTargetMatches(
                {
                    "targetId": "target-1",
                    "type": "service_worker",
                    "url": "chrome-extension://aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa/background.js",
                }
            )
        )

    def test_extensioninjector_base_inject_reports_the_subclass_name(self) -> None:
        with self.assertRaisesRegex(NotImplementedError, "ExtensionInjector.inject is not implemented"):
            ExtensionInjector().inject()


if __name__ == "__main__":
    unittest.main()

File: js/test/test.ExtensionInjector.ts
// MODCDP_TRANSLATE_TEST: KEEP THIS TEST FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
// All test cases, descriptions, covered edge cases, and setup should be kept perfectly 1:1 in sync between:
// - ./python/tests/test_ExtensionInjector.py
// - ./go/modcdp/injector/ExtensionInjector_test.go
// NO MOCKING, NO MONKEY PATCHING, NO SIMULATING, NO FAKING, NO SKIPPING ALLOWED.
// USE REAL USER-FACING CODE PATHS WITH REAL BROWSERS, REAL CLASSES, REAL URLS, etc. Hard fail if keys or other env requirements are missing.
import assert from "node:assert/strict";
import { test } from "vitest";

import { ExtensionInjector } from "../src/injector/ExtensionInjector.js";

class ProbeExtensionInjector extends ExtensionInjector {
  matches(target: { type?: string; url?: string }) {
    return this.serviceWorkerTargetMatches(target);
  }
}

test("ExtensionInjector owns shared injector config", async () => {
  const injector = new ProbeExtensionInjector({
    injector_service_worker_extension_id: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
    injector_service_worker_url_suffixes: ["/modcdp/service_worker.js"],
  });

  try {
    assert.equal(injector.config.injector_service_worker_extension_id, "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa");
    assert.deepEqual(injector.extra_args, []);
    assert.equal(
      injector.matches({
        type: "service_worker",
        url: "chrome-extension://aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa/modcdp/service_worker.js",
      }),
      true,
    );
    assert.equal(
      injector.matches({
        type: "service_worker",
        url: "chrome-extension://aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa/background.js",
      }),
      false,
    );
  } finally {
    await injector.close();
  }
});

test("ExtensionInjector base inject reports the subclass name", async () => {
  await assert.rejects(() => new ExtensionInjector().inject(), /ExtensionInjector\.inject is not implemented/);
});
